### ISSUE TITLE
What if duck types are unions of all the ducklings

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#11516]
If the duck typed types were actually unions of all the ducklings then it would fix a bunch of edge cases that arise.
```py
def func1(f: float):
    reveal_type(f)  # float | int | complex

def func2(b: bytes):
    reveal_type(b)  # bytes | bytearray | memoryview

def func3(u: unicode):  # python 2 moment
    reveal_type(u)  # unicode | str
```

edge cases:
```py
def func(f: float):
    f.is_integer()  # error: int has no member "is_integer"
```
```py
def func(f: float):
    f.is_integer()  # error: int has no member "is_integer"
```

```py
def func(f: float):
    if not isinstance(f, float):
        print("hi")  # no 'unreachable code' error
        reveal_type(f)  # int | complex
```

It would also be useful to have type types for `float` and `bytes`, when you want exactly `float` and don't want any stinking `int`s.

Related # 11511, #11145

**Comment by CarliJoy** [comment]
See also: https://github.com/python/mypy/issues/12824

**Comment by erictraut** [comment]
This feature request appears to be based on a misunderstanding of the Python type system. The type annotation `float` doesn't simply refer to the class `float`. It represents a set of types that includes `float` and all subtypes thereof. The type `float` and the type `float | int` refer to the same set of types, so they are equivalent.

**Comment by KotlinIsland** [comment]
> This feature request appears to be based on a misunderstanding of the Python type system.

That's obviously not what I am saying lol. This issue is regarding addressing the problems introduced by the 'ducktyping' functionality [defined in pep 484](https://peps.python.org/pep-0484/#the-numeric-tower), which states that `float`/`int` should lie about their base classes ([mypy extends this idea to bytes](https://mypy.readthedocs.io/en/stable/duck_type_compatibility.html)). The solution being that instead of lying about their bases (extremely confusing), these specific types should just expand to a union of the ducklings.

**Comment by erictraut** [comment]
I don't understand what you're asking for here. I'm also not sure that you understood my previous point, so let me try again.

In type theory, if you have a class `Parent` and a class `Child` that derives from parent, then the types `Parent` and `Parent | Child` are 100% equivalent. If a type checker ever treats these as different, then it's a bug. If I understand you correctly, you're saying that `float` should be "expanded to" `float | int`, but those two types are 100% equivalent already because `int` is (by definition in PEP 484) a subtype of `float`. If mypy ever treats `float` differently from `float | int`, then it's a bug.

Are you suggesting that mypy add a new mode where it deviates from the behavior specified in PEP 484 and treats `int` as though it's not a subtype of `float`, etc.? That would be incompatible with all existing type stubs, so I don't think it would be very useful.

Or perhaps what you're asking is for new special forms to be added to the type system that represent "a float that is not an int" (`float & ~int`), etc.? That would require a new specification in the form of a PEP, and the mypy issue tracker wouldn't be the right place to request such a change.

Or perhaps what you're saying is that when mypy prints the type `builtins.float` in a textual form (e.g. in error messages), you always want it to print `builtins.float | builtins.int`. That would be sound from a type perspective, but I think most mypy users would find that representation to be redundant, verbose, and even somewhat confusing.

Or perhaps I haven't yet captured what you're proposing. If that's the case, can you provide more specifics?

**Comment by ikonst** [comment]
As Eric said, `int | float` is equivalent to `float` in our current type system. So let's assume there's a form of an exact identity `Is[float]`, it could upcast to `float` but not vice versa, else it'd be easy to sabotage:
```python
def f(arg: Is[float]) -> None:
  assert isinstance(arg, float) # FAIL

a: float = 42
f(a)
```
so you'd have to either narrow with `type(x) is float` or pass `Is[float]`s around (a typical [red-blue](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) in the making ;)).

**Comment by ikonst** [comment]
And then there's the question of "why?". It might be marginally useful for debugging or when types violate the LSP.

**Comment by KotlinIsland** [comment]
@ikonst yeah, exactly. I plan on implementing this in basedmypy, it will be an interesting experiment. 
- kotlinisland/basedmypy#168

**Comment by erictraut** [comment]
@ikonst, what do you mean by "emit `float | int` behind the scenes"? Will that result in any user-visible behavior change in mypy? If so, how? Is the only visible change that the textual output for `reveal_type` and error messages becomes more verbose?

**Comment by KotlinIsland** [comment]
> @ikonst, what do you mean by "emit `float | int` behind the scenes"? Will that result in any user-visible behavior change in mypy? If so, how? Is the only visible change that the textual output for `reveal_type` and error messages becomes more verbose?

No, annotating something with `float` will result in the type being `float | int`. And the promotion/ducktyping will be removed from `int`.

**Comment by erictraut** [comment]
Ah, I see what you mean. Thanks for the explanation. Yeah, that's an interesting experiment. Let us know what you find.
